components/script/dom/idbobjectstore.rs

```diff
@@ -0,0 +1,589 @@
+    ) -> DomRoot<IDBObjectStore> {
+        reflect_dom_object(
+            Box::new(IDBObjectStore::new_inherited(
+                global, db_name, name, options,
+            )),
+            global,
+            IDBObjectStoreBinding::Wrap,
+        )
+    }
+
+    pub fn get_name(&self) -> DOMString {
+        self.name.borrow().clone()
+    }
+
+    pub fn set_transaction(&self, transaction: &IDBTransaction) {
+        self.transaction.set(Some(transaction));
+    }
+
+    pub fn transaction(&self) -> Option<DomRoot<IDBTransaction>> {
+        self.transaction.get()
+    }
+
+    // https://www.w3.org/TR/IndexedDB-2/#valid-key-path
+    pub fn is_valid_key_path(key_path: &StrOrStringSequence) -> bool {
+        let is_valid = |path: &DOMString| {
+            let path = path.to_string();
+            let mut identifiers = path.split('.').into_iter();
+
+            while let Some(_identifier) = identifiers.next() {
+                // FIXME:(rasviitanen) Implement this properly
+            }
+
+            true
+        };
+
+        match key_path {
+            StrOrStringSequence::StringSequence(paths) => {
+                if paths.is_empty() {
+                    return false;
+                }
+
+                for path in paths {
+                    if !is_valid(path) {
+                        return false;
+                    }
+                }
+                true
+            },
+            StrOrStringSequence::String(path) => is_valid(path),
+        }
+    }
+
+    #[allow(unsafe_code)]
+    // https://www.w3.org/TR/IndexedDB-2/#convert-value-to-key
+    fn convert_value_to_key(
+        cx: SafeJSContext,
+        input: HandleValue,
+        seen: Option<Vec<HandleValue>>,
+    ) -> Result<IndexedDBKeyType, Error> {
+        // Step 1: If seen was not given, then let seen be a new empty set.
+        let _seen = seen.unwrap_or(Vec::new());
+
+        // Step 2: If seen contains input, then return invalid.
+        // FIXME:(rasviitanen)
+        // Check if we have seen this key
+        // Does not currently work with HandleValue,
+        // as it does not implement PartialEq
+
+        // Step 3
+        // FIXME:(rasviitanen) Accept buffer, array and date as well
+        if input.is_number() {
+            // FIXME:(rasviitanen) check for NaN
+            let key = structuredclone::write(cx, input, None).expect("Could not serialize key");
+            return Ok(IndexedDBKeyType::Number(key.serialized.clone()));
+        }
+
+        if input.is_string() {
+            let key = structuredclone::write(cx, input, None).expect("Could not serialize key");
+            return Ok(IndexedDBKeyType::String(key.serialized.clone()));
+        }
+
+        if input.is_object() {
+            rooted!(in(*cx) let object = input.to_object());
+            unsafe {
+                let built_in_class: *mut ESClass =
+                    libc::malloc(mem::size_of::<ESClass>()) as *mut ESClass;
```

Comment:
Rather than using malloc, let's do `let mut built_in_class = ESClass::Other;` and pass `&mut built_in_class` to GetBuiltInClass.